File: app/src/main/java/io/github/jangerhard/BitcoinWalletTracker/model/BlockinfoResponse.java
package io.github.jangerhard.BitcoinWalletTracker.model;

import io.vavr.collection.List;

public class BlockinfoResponse {

    private String address;
    private long n_tx;
    private long total_received;
    private long total_sent;
    private long final_balance;
    private List<Transaction> txs;

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public long getN_tx() {
        return n_tx;
    }

    public void setN_tx(long n_tx) {
        this.n_tx = n_tx;
    }

    public long getTotal_received() {
        return total_received;
    }

    public void setTotal_received(long total_received) {
        this.total_received = total_received;
    }

    public long getTotal_sent() {
        return total_sent;
    }

    public void setTotal_sent(long total_sent) {
        this.total_sent = total_sent;
    }

    public long getFinal_balance() {
        return final_balance;
    }

    public void setFinal_balance(long final_balance) {
        this.final_balance = final_balance;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;

        BlockinfoResponse acc = (BlockinfoResponse) obj;

        return (address.equals(acc.getAddress()) &&
                n_tx == acc.n_tx &&
                total_received == acc.total_received &&
                total_sent == acc.total_sent &&
                final_balance == acc.final_balance);
    }

    @Override
    public String toString() {
        return address + " has a balance of " + getFinal_balance() + "btc.";
    }

    public List<Transaction> getTxs() {
        return txs;
    }

    public void setTxs(List<Transaction> txs) {
        this.txs = txs;
    }
}

File: app/src/main/java/io/github/jangerhard/BitcoinWalletTracker/model/BlockinfoResponse.kt
package io.github.jangerhard.BitcoinWalletTracker.model

import io.vavr.collection.List

data class BlockinfoResponse(
        val address: String,
        val n_tx: Long,
        val total_received: Long,
        val total_sent: Long,
        val final_balance: Long,
        val txs: List<Transaction> = List.empty()
) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as BlockinfoResponse

        if (address != other.address) return false
        if (n_tx != other.n_tx) return false

        return true
    }

    override fun hashCode(): Int {
        var result = address.hashCode()
        result = 31 * result + n_tx.hashCode()
        return result
    }

    override fun toString(): String {
        return "BlockinfoResponse(address='$address', n_tx=$n_tx, total_received=$total_received, total_sent=$total_sent, final_balance=$final_balance, txs=$txs)"
    }

    data class Transaction(
            val result: Long,
            val out: List<TransactionOut> = List.empty(),
            val inputs: List<TransactionInput> = List.empty(),
            val time: Long
    ) {

        data class TransactionOut(
                val spent: Boolean,
                val value: Long,
                val addr: String
        )

        data class TransactionInput(
                val prevOut: TransactionOut?
        )


        override fun toString(): String {
            return "Transaction(result=$result, out=$out, inputs=$inputs, time=$time)"
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Transaction

            if (result != other.result) return false
            if (out != other.out) return false
            if (inputs != other.inputs) return false
            if (time != other.time) return false

            return true
        }

        override fun hashCode(): Int {
            var result1 = result.hashCode()
            result1 = 31 * result1 + out.hashCode()
            result1 = 31 * result1 + inputs.hashCode()
            result1 = 31 * result1 + time.hashCode()
            return result1
        }

    }

}
